diff --git ircd/s_serv.c ircd/s_serv.c
index 571ff60..ccb6082 100644
--- ircd/s_serv.c
+++ ircd/s_serv.c
@@ -768,9 +768,10 @@ int m_server(aClient *cptr, aClient *sptr, int parc, char *parv[])
             IsIPv6(acptr) ? "6" : "", PunteroACadena(acptr->info));
 #if !defined(NO_PROTOCOL9)
       else
-        sendto_one(bcptr, ":%s SERVER %s %d 0 %s %s %s%s 0 :%s",
+        sendto_one(bcptr, ":%s SERVER %s %d 0 %s %s %s%s +%s%s%s :%s",
             parv[0], PunteroACadena(acptr->name), hop + 1, parv[4], parv[5],
-            NumServCap(acptr), PunteroACadena(acptr->info));
+            NumServCap(acptr), IsHub(acptr) ? "h" : "", IsService(acptr) ? "s" : "",
+            IsIPv6(acptr) ? "6" : "", PunteroACadena(acptr->info));
 #endif
     }
     return 0;
@@ -946,9 +947,14 @@ int m_server_estab(aClient *cptr, aConfItem *aconf, aConfItem *bconf, time_t sta
           me.info ? me.info : "IRCers United");
     else
       sendto_one(cptr,
-          "SERVER %s 1 " TIME_T_FMT " " TIME_T_FMT " J%s %s%s 0 :%s",
+          "SERVER %s 1 " TIME_T_FMT " " TIME_T_FMT " J%s %s%s +6%s :%s",
            my_name_for_link(me.name, aconf), me.serv->timestamp,
-           cptr->serv->timestamp, MAJOR_PROTOCOL, NumServCap(&me), 
+           cptr->serv->timestamp, MAJOR_PROTOCOL, NumServCap(&me),
+#if defined(HUB)
+          "h",
+#else
+          "",
+#endif
            me.info ? me.info : "IRCers United");
 
     tx_num_serie_dbs(cptr);
@@ -1029,9 +1035,10 @@ int m_server_estab(aClient *cptr, aConfItem *aconf, aConfItem *bconf, time_t sta
 #if !defined(NO_PROTOCOL9)
       else
         sendto_one(acptr,
-            ":%s SERVER %s 2 0 " TIME_T_FMT " %s%u %s%s 0 :%s", me.name,
+            ":%s SERVER %s 2 0 " TIME_T_FMT " %s%u %s%s +%s%s%s :%s", me.name,
             cptr->name, cptr->serv->timestamp,
             (Protocol(cptr) > 9) ? "J" : "J0", Protocol(cptr), NumServCap(cptr),
+            IsHub(cptr) ? "h" : "", IsService(cptr) ? "s" : "", IsIPv6(cptr) ? "6" : "",
             PunteroACadena(cptr->info));
 #endif
     }
@@ -1045,10 +1052,11 @@ int m_server_estab(aClient *cptr, aConfItem *aconf, aConfItem *bconf, time_t sta
             IsIPv6(cptr) ? "6" : "", PunteroACadena(cptr->info));
 #if !defined(NO_PROTOCOL9)
       else
-        sendto_one(acptr, ":%s SERVER %s 2 0 " TIME_T_FMT " %s%u %s%s 0 :%s",
+        sendto_one(acptr, ":%s SERVER %s 2 0 " TIME_T_FMT " %s%u %s%s +%s%s%s :%s",
             me.name, cptr->name, cptr->serv->timestamp,
             (Protocol(cptr) > 9) ? "J" : "J0", Protocol(cptr),
-            NumServCap(cptr), PunteroACadena(cptr->info));
+            NumServCap(cptr), IsHub(cptr) ? "h" : "", IsService(cptr) ? "s" : "",
+            IsIPv6(cptr) ? "6" : "", PunteroACadena(cptr->info));
 #endif
     }
   }
@@ -1095,10 +1103,11 @@ int m_server_estab(aClient *cptr, aConfItem *aconf, aConfItem *bconf, time_t sta
               PunteroACadena(acptr->info));
         else
           sendto_one(cptr,
-              ":%s SERVER %s %d 0 " TIME_T_FMT " %s%u %s%s 0 :%s",
+              ":%s SERVER %s %d 0 " TIME_T_FMT " %s%u %s%s +%s%s%s :%s",
               acptr->serv->up->name, acptr->name,
               acptr->hopcount + 1, acptr->serv->timestamp,
               protocol_str, Protocol(acptr), NumServCap(acptr),
+              IsHub(acptr) ? "h" : "", IsService(acptr) ? "s" : "", IsIPv6(acptr) ? "6" : "",
               PunteroACadena(acptr->info));
       }
       else
@@ -1114,10 +1123,11 @@ int m_server_estab(aClient *cptr, aConfItem *aconf, aConfItem *bconf, time_t sta
 #if !defined(NO_PROTOCOL9)
         else
           sendto_one(cptr,
-              ":%s SERVER %s %d 0 " TIME_T_FMT " %s%u %s%s 0 :%s",
+              ":%s SERVER %s %d 0 " TIME_T_FMT " %s%u %s%s +%s%s%s :%s",
               acptr->serv->up->name, acptr->name,
               acptr->hopcount + 1, acptr->serv->timestamp,
               protocol_str, Protocol(acptr), NumServCap(acptr),
+              IsHub(acptr) ? "h" : "", IsService(acptr) ? "s" : "", IsIPv6(acptr) ? "6" : "",
               PunteroACadena(acptr->info));
 #endif
       }
@@ -1139,11 +1149,15 @@ int m_server_estab(aClient *cptr, aConfItem *aconf, aConfItem *bconf, time_t sta
          * been received. -avalon
          * Or only NICK in new format. --Run
          */
-        sendto_one(cptr, ":%s NICK %s %d " TIME_T_FMT " %s %s %s :%s",
+        char xxx_buf[25];
+
+        sendto_one(cptr, ":%s NICK %s %d " TIME_T_FMT " %s %s %s %s %s%s :%s",
             acptr->user->server->name,
             acptr->name, acptr->hopcount + 1, acptr->lastnick,
             PunteroACadena(acptr->user->username),
             PunteroACadena(acptr->user->host), acptr->user->server->name,
+            iptobase64(xxx_buf, &acptr->ip, sizeof(xxx_buf), 1),
+            NumNick(acptr),
             PunteroACadena(acptr->info));
 
         send_umode(cptr, acptr, 0, SEND_UMODES, 0, SEND_HMODES);
diff --git a/ircd/s_user.c b/ircd/s_user.c
index 0755a8f..90130e7 100644
--- ircd/s_user.c
+++ ircd/s_user.c
@@ -820,10 +820,12 @@ static int register_user(aClient *cptr, aClient *sptr,
 #else /* Remove the following when all servers are 2.10 */
 
   /* First send message to all 2.9 servers */
-  sprintf_irc(sendbuf, ":%s NICK %s %d " TIME_T_FMT " %s %s %s :%s",
+  sprintf_irc(sendbuf, ":%s NICK %s %d " TIME_T_FMT " %s %s %s %s %s%s :%s",
       user->server->name, nick, sptr->hopcount + 1, sptr->lastnick,
       PunteroACadena(user->username), PunteroACadena(user->host),
-      user->server->name, PunteroACadena(sptr->info));
+      user->server->name, iptobase64(ip_base64, &sptr->ip, sizeof(ip_base64), 1),
+      NumNick(sptr),
+      PunteroACadena(sptr->info));
 
   for (lp = me.serv->down; lp; lp = lp->next)
   {
