index 27a41c8..9f92268 100644
--- ircd/send.c
+++ ircd/send.c
@@ -412,7 +412,7 @@ static void vsendto_prefix_one(aClient *to, aClient *from,
       {
         strcat(sender, "@");
 #if defined(BDD_VIP)
-        strcat(sender, get_visiblehost(from, NULL, 0));
+        strcat(sender, get_visiblehost(from, to, 0));
 #else
         strcat(sender, user->host);
 #endif
@@ -427,7 +427,7 @@ static void vsendto_prefix_one(aClient *to, aClient *from,
     {
       strcat(sender, "@");
 #if defined(BDD_VIP)
-      strcat(sender, get_visiblehost(from, NULL, 0));
+      strcat(sender, get_visiblehost(from, to, 0));
 #else
       if (IsUnixSocket(from))
         strcat(sender, user->host);
